bind/gen.go

```diff
@@ -0,0 +1,666 @@
+// the following are used for build value -- switch on reflect.Kind
+// or the types equivalent
+static inline PyObject* gopy_build_bool(uint8_t val) {
+	return Py_BuildValue("b", val);
+}
+static inline PyObject* gopy_build_int64(int64_t val) {
+	return Py_BuildValue("k", val);
+}
+static inline PyObject* gopy_build_uint64(uint64_t val) {
+	return Py_BuildValue("K", val);
+}
+static inline PyObject* gopy_build_float64(double val) {
+	return Py_BuildValue("d", val);
+}
+static inline PyObject* gopy_build_string(const char* val) {
+	return Py_BuildValue("s", val);
+}
+static inline void gopy_decref(PyObject* obj) { // macro
+	Py_XDECREF(obj);
+}
+static inline void gopy_incref(PyObject* obj) { // macro
+	Py_XINCREF(obj);
+}
+static inline int gopy_method_check(PyObject* obj) { // macro
+	return PyMethod_Check(obj);
+}
+static inline void gopy_err_handle() {
+	if(PyErr_Occurred() != NULL) {
+		PyErr_Print();
+	}
+}
+%[8]s
+*/
+import "C"
+import (
+	"github.com/goki/gopy/gopyh" // handler
+	%[6]s
+)
+
+// main doesn't do anything in lib / pkg mode, but is essential for exe mode
+func main() {
+	%[7]s
+}
+
+// initialization functions -- can be called from python after library is loaded
+// GoPyInitRunFile runs a separate python file -- call in GoPyInit if it
+// steals the main thread e.g., for GUI event loop, as in GoGi startup.
+
+//export GoPyInit
+func GoPyInit() {
+	%[7]s
+}
+
+// type for the handle -- int64 for speed (can switch to string)
+type GoHandle %[4]s
+type CGoHandle %[5]s
+
+// boolGoToPy converts a Go bool to python-compatible C.char
+func boolGoToPy(b bool) C.char {
+	if b {
+		return 1
+	}
+	return 0
+}
+
+// boolPyToGo converts a python-compatible C.Char to Go bool
+func boolPyToGo(b C.char) bool {
+	if b != 0 {
+		return true
+	}
+	return false
+}
+
+func complex64GoToPy(c complex64) *C.PyObject {
+	return C.PyComplex_FromDoubles(C.double(real(c)), C.double(imag(c)))
+}
+
+func complex64PyToGo(o *C.PyObject) complex64 {
+	v := C.PyComplex_AsCComplex(o)
+	return complex(float32(v.real), float32(v.imag))
+}
+
+func complex128GoToPy(c complex128) *C.PyObject {
+	return C.PyComplex_FromDoubles(C.double(real(c)), C.double(imag(c)))
+}
+
+func complex128PyToGo(o *C.PyObject) complex128 {
+	v := C.PyComplex_AsCComplex(o)
+	return complex(float64(v.real), float64(v.imag))
+}
+
+%[9]s
+`
+
+	goExePreambleC = `
+#if PY_VERSION_HEX >= 0x03000000
+extern PyObject* PyInit__%[1]s(void);
+static inline void gopy_load_mod() {
+	PyImport_AppendInittab("_%[1]s", PyInit__%[1]s);
+}
+#else
+extern void* init__%[1]s(void);
+static inline void gopy_load_mod() {
+	PyImport_AppendInittab("_%[1]s", init__%[1]s);
+}
+#endif
+
+`
+
+	goExePreambleGo = `
+// wchar version of startup args
+var wargs []*C.wchar_t
+
+//export GoPyMainRun
+func GoPyMainRun() {
+	// need to encode char* into wchar_t*
+	for i := range os.Args {
+		wargs = append(wargs, C.Py_DecodeLocale(C.CString(os.Args[i]), nil))
```

Comment:
what's the lifetime of the c-string created with `C.CString` here?
do we need to `C.free` it?
this probably needs a comment.

Reply:
added issue..